Add page about Scala 2 `-Xsource:3` to migration guide (#2966)
scala/docs.scala-lang: _overviews/scala3-migration/tooling-scala2-xsource3.md

---
title: Scala 2 with -Xsource:3
type: chapter
description: This section describes the Scala 2 compiler's -Xsource:3 flag
num: 7
previous-page: tooling-tour
next-page: tooling-migration-mode
---

The Scala 2.13 compiler issues helpful migration warnings with the `-Xsource:3` flag.
Before moving to the Scala 3 compiler, it's recommended to enable this flag in Scala 2 and address the new warnings.

> In addition to new warnigns, the flag enables certain benign Scala 3 syntax such as `import p.*`.

Reviewer: ```suggestion
In addition to new warnings, the flag enables certain benign Scala 3 syntaxes such as `import p.*`.
```

This page explains the details behind the flag, an overview is shown using `scalac -Xsource:help`.

## Migration vs cross-building

With Scala 2.13.13 and newer, the `-Xsource:3` flag comes in two variants:

>   - `Xsource:3` enables warnings relevant for migrating a codebase to Scala 3.

Reviewer: Maybe add something like:

> It also enables a few benign Scala 3 syntaxes such as `import foo.*`.

>   - `Xsoruce:3-cross` is useful for projects that cross-build between Scala 2 and 3 for a longer period of time.

Reviewer 2: ```suggestion
  - `Xsource:3-cross` is useful for projects that cross-build between Scala 2 and 3 for a longer period of time.
```

>     For certain language constructs that trigger a warning with `-Xsoruce:3`, the behavior changes to match Scala 3.

Reviewer 2: ```suggestion
    For certain language constructs that trigger a warning with `-Xsource:3`, the behavior changes to match Scala 3.
```

Details about individual warnings are listed below on this page.

## Fatal warnings and quick fixes

By default, Scala 3 migration warnings emitted by Scala 2.13 are fatal, i.e., they are reported as errors.
This can be changed using `-Wconf`, for example `-Wconf:cat=scala3-migration:w` changes them to be reported as warnings.
Alternatively, `-Xmigration` has the same effect.

The [`@nowarn` annotation](https://www.scala-lang.org/api/current/scala/annotation/nowarn.html) can be used to suppress individual warnings, which also works with fatal warnings enabled.

The Scala 2.13 compiler implements quick fixes for many Scala 3 migration warnings.

> Quick fixes are displayed in Metals-based IDEs (not yet in IntelliJ), and they can be applied directly to the source code using the `-quickfix` flag, for example `-quickfix:cat=scala3-migration`.

Reviewer: Not yet in IntelliJ? Really? Is there a ticket somewhere tracking that?

Author: Not sure, if we find one we should add it to https://github.com/lampepfl/dotty/issues/14904
